"always RVF" arrays and provide an option to serialize certain slices

bradcray opened the pull request:
This change makes the compiler "always RVF" the `_array` record and provides serialize/deserialize routines for many common domain maps.  It also provides an option to remote-value forward slices via an (intentionally) undocumented config param named `chpl_serializeSlices`, currently defaulted to "off" such that the status quo for how slices are communicated will not change in this PR.

By default, we currently proactively communicate with all locales to create a privatized version of an array slice descriptor when slicing a privatized (essentially, distributed) array.  This is inherently expensive, and particularly pointless in cases where the slice only covers a small number of locales.  The original point of this PR was to reduce these costs by RVF-ing such slices; however, it turned out that the approach was not always a win, so currently it's an opt-in capability until we can make it more of a win by default.

When `chpl_serializeSlices` is set to `true`, this PR attempts to RVF slices across locales as the compiler encounters on-clauses.  For such cases, creating slices of distributed domains becomes communication-free, and any subsequent communication of the slice takes place after-the-fact, either by bundling the slice into an on-clause using RVF or by communicating to get it otherwise.  For common cases, like slicing a block-distributed array with a block-distributed domain, the serialization of the slice passes a tuple of the array and domain PIDs, permitting the receiving side to look up the corresponding privatized classes on its side without additional communication.

The previous paragraph says "attempts" and "otherwise" because (a) RVFing doesn't fire for all arrays that span an on-clause and (b) this capability hasn't been enabled for all distributions in this PR.  W.r.t. (a), global slices are not currently RVF'd, nor are slices that are part of a zippering expression.  W.r.t. (b), slices query the array and domain involved in the slice operation to see whether they can be serialized and only indicate that the slice can be serialized if those components can be.  In this PR, I've implemented serialization for several common distributions, but not all of them.

In more detail, the approach taken here is to apply the "always RVF" pragma to the _array type (which is arguably what the compiler was already doing via a special-case for arrays).  It then applies a where-clause to the `serialize()` / `deserialize()` overloads on the _array record and domain map classes to only have serialization fire in cases where we want them to (kudos to @benharsh for suggesting this where-clause based approach).

Specifically, the where clause on arrays is based on a new `chpl__rvfMe()` routine defined for arrays which says whether or not to use the serialize routines to RVF a given array implementation.  This is defined to be `false` by default in the base class which causes us to fall back on the compiler's typical bit copy implementation.  We then overload it for slices (ArrayViewSliceArr) to return true when the domain and array are each privatized and each has a serialize() routines defined for it.  

The serialize/deserialize routines for the _array record are now rewritten to recursively call into the wrapped class's serialize/deserialize routines rather than being special-cased as before.  For slices, the serialize/deserialize routines simply calls the serialize()/deserialize() routines on the slicing domain and sliced array respectively, returning the result as a tuple.

We then add serialize/deserialize routines for key privatized distributions (Block, Stencil, Cyclic, BlockCyclic) that simply act in terms of their PIDs.  As a result a Block array sliced by a Block domain is RVF'd by sending a tuple of the two PIDs (one for the array, one for the domain).

Other work involved included:
* optimizing leader/follower code for array views to avoid unnecessary communications
* disabling forwarding of the serialize/deserialize calls for array views
* making privatization support for slices conditional on the `chpl_serializeSlices` config param
* moving and cleaning up code relating to serializing domains into DefaultRectangular.chpl since it's the only case using it and this work suggests that Block, Cyclic, etc. don't need to build on that support as originally implied by the comments
* tightening up a special-case related to RVF in the compiler that was introduced for domains to make sure it doesn't get applied to arrays

In terms of testing:
* test/distributions/robust/arithmentic/performance/multilocale/sliceOps.*: added a new test to measure various key slicing idioms that were helpful in determining where things are improved vs. not
* test/distributions/robust/arithmentic/performance/multilocale/rvfSlices/*: added this subdirectory to run the four main existing slicing tests with this new serialization option on (as well as a verbose debugging mode) to lock in the current behavior and comm counts
* test/arrays/slices/commCounts/*: added new tests that count the amount of communication used under this PR when slicing a Block array by a Block, DR, or tuple of ranges; as well as for a Block slice of a DR array.


Future work:
- [ ] once we can determine that a domain is const, we should also RVF const DR domains used to slice

bradcray commented:
This also improves the comm counts in the exchange test:

exchange-elegant.dat:
```
05/10/19        1634    104     540 # master
05/10/19        1623    105     547
05/10/19        1604    102     516
05/10/19        1616    103     524
05/10/19        1641    107     531

05/10/19        1595    88      358  # put-rvf-slice-on-master
05/10/19        1565    87      357
05/10/19        1610    85      369
05/10/19        1632    90      367
05/10/19        1633    90      364
```

exchange-destLocal.dat:
```
05/10/19        780     60      281  # master
05/10/19        811     63      290
05/10/19        804     64      271
05/10/19        819     64      290
05/10/19        783     60      285

05/10/19        803     54      202  # put-rvf-slice-on-master
05/10/19        761     51      197
05/10/19        783     52      203
05/10/19        774     47      200
05/10/19        763     47      198
```

though I expect the impact to be bigger once we are no longer creating block-distributed domains whenever we use ranges to slice block-distributed arrays.

bradcray commented:
@benharsh: I think that this is ready for review and that you have the most relevant expertise for this PR.  Would you be able to give it a review for me?  I'm going to look a bit more into the cyclic comm count regressions in the meantime.

benharsh commented on modules/internal/ArrayViewSlice.chpl:
```diff
@@ -78,6 +78,50 @@ module ArrayViewSlice {
+    // Serialize a slice (when chpl__rvfMe() says to) by serializing its
+    // domain and array
+    //
+    proc chpl__serialize() where chpl__rvfMe() {
```
You might want to disable forwarding for chpl__serialize/deserialize for this ArrayView as well because it can forward if the where-clause is 'false'.

benharsh commented on modules/internal/ChapelArray.chpl:
```diff
@@ -2205,6 +2171,26 @@ module ChapelArray {
+    }
+
+    proc type chpl__typeOfInstance() type {
+      var x: this;
```
Can you not use the "static field type" primitive here as well?

bradcray replied:
I was wondering about unifying these when setting up this PR as well, but figured I'd get it reviewed first in case I had to do other edits.  Do you prefer using the primitive vs. the helper routine?  (tradeoff seems to be "relies on a primitive" vs. "not"?)

benharsh replied:
Mild preference for the primitive, but I wouldn't hold up the PR over it. I was more curious whether there was a meaningful difference.

bradcray commented:
> could these tests call into some common utility module instead of duplicating the same code?

Which tests are you referring to here?

benharsh commented:
The new ones under ``test/arrays/slices/commCounts`` seemed to share the same general format, just with different slices (maybe I read too quickly?).

bradcray commented:
Oh right, sorry...  there'd only been one new tests for the months I developed this and I added the other variants at the last minute on Friday and then forgot that they existed.

bradcray commented:
OK, I figured out what's going wrong with the Cyclic comm counts (why the number of gets went up for Cyclic only and not others):  The main problem was that my mental model for how RVFing works continuens to be too naive.  In my mind, if I saw "always RVF" on something, I imagine that any reference to that something within an on-clause will be local as a result of the RVF.  But this isn't always the case.  For example, we never RVF global variables even if they say "always RVF" (and I think there are other cases where the on-clause is sufficiently far from the array that we might also not RVF, but I need to verify that still...).  And then, unfortunately, to compound the problem, if I saw the comms for creating a slice listed as "none", that confirmed my sense that we were RVFing the slice when in fact all it meant was that we weren't privatizing it.

So... what was happening with the assignSlice case, for example, is that the slice was global so would not be RVF'd, but the comm counts would drop as expected.  Meanwhile, the block case did not generate extra gets, suggesting to me that it was being RVF'd.  In fact what was happening was that it was going down the bulk transfer path, whereas Cyclic does not (bulk transfer isn't supported on Cyclic yet), so it would enter a forall loop, and the slice's follower would run `proc arr()` in the array slice code for each iteration.  And since RVF hadn't fired, this would refer back to the original slice descriptor on locale 0 in each iteration.  This turned out to be harder to diagnose than I expected because the number of iterations went up by ~20 or so, which is similar to the magnitudes I was seeing for things like "any iteration over a Cyclic array", so I thought it was an O(1) cost.  It turns out it was really an O(n) cost and that the assignSlice program just uses a small enough problem size that it didn't result in many gets.

So a workaround for this is to hoist the call to `proc arr()` out of the follower loop at which point we do O(numTasks) gets for each forall loop rather than O(n) times as in the previous versions of my branch.  But still more than the 0 gets that we did when we aggressively privatized.  This makes me worry that RVFing slices is less of an obvious no-brainer than I was originally thinking.

This also leads to questions about whether we could/should be doing more aggressive RVFing or perhaps thinking of this "bundle slices with on-clauses" transformation in terms of some other mechanism than RVF.   It also seems somewhat similar to issue #10160 in that the slice descriptor is essentially a class whose field is (logically) a privatized array and for which we'd like to avoid communication back to the originating array all the time.

Anyway, that's the mystery of what was going on for the Cyclic distribution explained, if not exactly solved.

bradcray commented:
Capturing some rough stats for today's performance meeting:

On master:
* cost of slicing block with block:
  - on: numLocales to privatize slice descriptor
* cost of using slice:
  - free, other than on-clauses required to fire off distributed computation

On branch, when RVF fires:
* cost of slicing block with block:
  - free
* cost of using slice:
  - free (data RVF'd with required tasks)

On branch, when RVF doesn't fire:
* cost of slicing block with block:
  - free
* cost of using slice
  - 2*numTasks gets (== numLocales that the slice covers * numCores per locale)

bradcray commented:
@benharsh: This has changed pretty substantially since your review, primarily to keep the new feature off by default given the better understanding of the comm count implications, and in response to your feedback:

* a config param has been added to toggle the feature on and off, defaulting to "off"
* another has been added to print when serialization is occurring to better distinguish between optimization due to serialization vs. due to bulk copy (since the resulting comm counts are often identical)
* I've restored all the comm count tests to their original values due the feature being off and added clones of them with both config params set to lock in the current behavior
* I reduced the new tests to two: one slicing by a domain and the other by ranges
* I switched to using the primitive to get the type of `_instance` in both cases
* I turned of forwarding of serialize/deserialize routines on slices
* I made the adjustment to the compiler's RVF pass that we discussed so that its ability to RVF _array records wasn't dependent on how the array was def'd / use'd and to detangle it from the domain's special case

Let me know if you still approve this PR.

bradcray commented:
As of PR #13092, the table for communications is now (where the change is to the last line; and where _branch_ now refers to master with `-schpl_serializeSlices=true`):

| scenario          | ONs | GETs  | potential mitigation strategy |
| --------------- | ----- | ----- |-----|
| create slice (master) | numLocales | none | |
| create slice (branch) | none | none | celebrate!! |
| standalone iter (master) | minimal | none | |
| standalone iter (branch) | minimal | none | |
| leader/follower (master) | minimal | none | |
| leader/follower (branch) | minimal | 2*numTasks | have compiler propagate serialization to L-F |
| leader/follower (master) | minimal | none | |
| leader/follower (branch) | minimal | 2*numTasks | the above, or a variation on it? |
| assign w/ bulk (master) | minimal | none | |
| assign w/ bulk (branch) | minimal | none | |
| assign w/out bulk (master) | minimal | none | |
| assign w/out bulk (branch) | minimal | 2*numTasks | implement bulk assignment for more types and/or the previous strategy |
| random access local (master) | minimal | none | |
| random access local (branch) | minimal | none | |
| random access global (master) | minimal | n | |
| random access global (branch) | minimal | **none** | |
